File: part1/scraper.py
#!/usr/bin/python3

import socket
import sys
import os


# try:
#     name = sys.argv[1]
#     if sys.argv[2].lower() == "yes" or sys.argv[2].lower() == "no":
#         response =  "1" if sys.argv[2] == "yes" else "0" #argv[2] == "yes" ? "1" : "0"
#     else:
#         print(f"Bad format\nExample usage:\n{str(os.path.basename(__file__))}  Rick yes\n{str(os.path.basename(__file__))}  Morty no")
#         exit(1)
# except IndexError:
#     print(f"Bad format\nExample usage:\n{str(os.path.basename(__file__))}  Rick yes\n{str(os.path.basename(__file__))}  Morty no")
#     exit(1)



name = "gdfsg"
response = "1"

# create an INET, STREAMing socket
s = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
# now connect to the web server on port 80 - the normal http port
s.connect(("www-test.cs.umanitoba.ca", 80))

# build request string
req = b""
req += b"POST /~wus2/cgi-bin/a.cgi HTTP/1.1\r\n"
req += b"Host: www-test.cs.umanitoba.ca\r\n"
req += str.encode(f"Content-Length: {14 + len(name)}\r\n")
req += b"Content-Type: application/x-www-form-urlencoded\r\n\r\n"

req += str.encode(f"name={name}&accept={response}")

# print(req.decode())
# print()
# print()
# print()

# Example POST request coming out from Chrome
################################################################################
# POST /~wus2/cgi-bin/a.cgi HTTP/1.1\r\n
# Host: www-test.cs.umanitoba.ca
# Connection: keep-alive
# Content-Length: 19
# Cache-Control: max-age=0
# Origin: http://www-test.cs.umanitoba.ca
# Upgrade-Insecure-Requests: 1
# DNT: 1
# Content-Type: application/x-www-form-urlencoded
# User-Agent: Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_6) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/84.0.4147.89 Safari/537.36
# Accept: text/html,application/xhtml+xml,application/xml;q=0.9,image/webp,image/apng,*/*;q=0.8,application/signed-exchange;v=b3;q=0.9
# Referer: http://www-test.cs.umanitoba.ca/~wus2/cgi-bin/a.cgi?Anonymize=
# Accept-Encoding: gzip, deflate
# Accept-Language: en-CA,en;q=0.9,zh-CN;q=0.8,zh;q=0.7,ja-JP;q=0.6,ja;q=0.5
# Cookie: name=; accept=
# 
# name=fdsaf&accept=1
################################################################################

# Debloated POST request to send
################################################################################
# POST /~wus2/cgi-bin/a.cgi HTTP/1.1
# Host: www-test.cs.umanitoba.ca
# Content-Length: 28
# Content-Type: application/x-www-form-urlencoded
#
# name=idsjuafabfbadf&accept=1
################################################################################
"""
/// @brief A test fuction to check if the server returned the thing we want.
///        It compares the result to check line by line, against a semi-hardcoded string.
/// @param toCheck string to check
/// @return `True` if check passed, `False` if failed.
///         If `-O` is not passed to interpreter(i.e. assersion is on), and a check failed, the function will `assert False` instead.
"""
def check(toCheck, name, response, comeback):
    # assert False is used intentionally to save typing.
    # the failing condition is already describe in the if condition
    if toCheck.find("chunked") == -1 and toCheck.find("Content-Length") != -1:
        print("not chunked, test not covered")
        assert False
        return False

    comeback_offset = 2 if comeback else 0

    # Build reference string
    REFERENCE = []
    REFERENCE.append("HTTP/1.1 200 OK")
    # skip Date
    # skip Server
    if not comeback:
        REFERENCE.append("Set-Cookie: name=" + name)
        REFERENCE.append("Set-Cookie: accept=" + response)
    # skip transfer encoding
    REFERENCE.append("Content-Type: text/html; charset=UTF-8")
    # skip 3 lines
    REFERENCE.append("""    <html>""")
    REFERENCE.append("""    <body>""")
    if comeback:
        REFERENCE.append("""<p style="color:green">Thanks for replying. Here's your previous response.</p>""")
    REFERENCE.append("""    <h1>You are Invited!</h1> """)
    REFERENCE.append("""    details goes here""")
    REFERENCE.append("""    <br/><br/><br/>""")
    REFERENCE.append("""    <form action="/~wus2/cgi-bin/a.cgi" method="POST">""")
    REFERENCE.append("""        <label for="name">Your name:</label>""")
    REFERENCE.append( f"        <input type=\"text\" name=\"name\" pattern=\".*[a-zA-Z].*\" title=\"At least 1 English letter\" value={name} autofocus required>")
    REFERENCE.append("""        <br/>""")
    REFERENCE.append("""        <p2>Would you accept the invitation?</p2>""")
    REFERENCE.append("""        <br/>""")
    if response == "0":
        REFERENCE.append("""        <input type="radio" name="accept" value="1" required > Accept""")
        REFERENCE.append("""        <input type="radio" name="accept" value="0" checked> Decline""")
    else:
        REFERENCE.append("""        <input type="radio" name="accept" value="1" required checked> Accept""")
        REFERENCE.append("""        <input type="radio" name="accept" value="0" > Decline""")
    REFERENCE.append("""        <br/>""")
    REFERENCE.append("""        <input type="submit" value="Submit">""")
    REFERENCE.append("""    </form>""")
    REFERENCE.append("""    <br/>""")
    REFERENCE.append("""        <form action="/~wus2/cgi-bin/a.cgi" method="GET">""")
    REFERENCE.append("""        <button name="Anonymize" type="submit">Anonymize</button>""")
    REFERENCE.append("""        </form>""")
    REFERENCE.append("""</body></html>""")

    # debloat responses from server
    processed = []
    toCheck = toCheck.splitlines(keepends = False)
    processed.append(toCheck[0])
    # skip Date
    # skip Server
    if not comeback:
        processed.append(toCheck[3]) # cookie
        processed.append(toCheck[4]) # cookie
    # skip transfer encoding
    processed.append(toCheck[6 - comeback_offset])   # content-type
    # skip 3 lines
    processed.append(toCheck[10 - comeback_offset])  # html
    processed.append(toCheck[11 - comeback_offset])  # body

    if comeback:
        processed.append(toCheck[10])
    elif  ((toCheck[12] != """<p style="color:green">Your reply has been submitted!</p>""") and 
            (toCheck[12] != """<p style="color:green">Your reply has been updated!</p>""")): 
        # print("FAIL 1")
        # print(toCheck[12])
        assert False
        return False


    # debloat HTML
    for i in range(13 - comeback_offset, len(toCheck)):
        if toCheck[i] == "" or toCheck[i].isspace():
            continue
        else:
            processed.append(toCheck[i])

    if (len(processed) != len(REFERENCE)):
        # print(processed, sep = '\n')
        # print("========================================================")
        # print(REFERENCE, sep = '\n')
        # print("FAIL 2")
        assert False
        return False

    for i in range(0, len(REFERENCE)):
        if (processed[i] != REFERENCE[i]):
            # print("FAIL 3")
            # print(processed[i] + " != " + REFERENCE[i])
            assert False
            return False

    # print("PASS")
    return True

# sample response from server
################################################################################
# HTTP/1.1 200 OK
# Date: Sat, 18 Jul 2020 21:10:34 GMT
# Server: Apache/2.4.6 (Scientific Linux) OpenSSL/1.0.2k-fips PHP/5.4.16 mod_wsgi/3.4 Python/2.7.5
# Set-Cookie: name=idsjuafabfadf
# Set-Cookie: accept=1
# Transfer-Encoding: chunked
# Content-Type: text/html; charset=UTF-8
#
# 35a
#
#     <html>
#     <body>
# <p style="color:green">Your reply has been updated!</p>
#
#     <h1>You are Invited!</h1> 
#     details goes here
#     <br/><br/><br/>
#
#     <form action="/~wus2/cgi-bin/a.cgi" method="POST">
#         <label for="name">Your name:</label>
#         <input type="text" name="name" pattern=".*[a-zA-Z].*" title="At least 1 English letter" value=idsjuafabfadf autofocus required>
#         <br/>
#         <p2>Would you accept the invitation?</p2>
#         <br/>
#         <input type="radio" name="accept" value="1" required checked> Accept
#         <input type="radio" name="accept" value="0" > Decline
#         <br/>
#         <input type="submit" value="Submit">
#     </form>
#     <br/>
#
#
#         <form action="/~wus2/cgi-bin/a.cgi" method="GET">
#         <button name="Anonymize" type="submit">Anonymize</button>
#         </form>
#
# </body></html>
################################################################################

# What if something went wrong
################################################################################
# HTTP/1.1 400 Bad Request
# Date: Sat, 18 Jul 2020 21:02:17 GMT
# Server: Apache/2.4.6 (Scientific Linux) OpenSSL/1.0.2k-fips PHP/5.4.16 mod_wsgi/3.4 Python/2.7.5
# Connection: close
# Content-Type: text/html; charset=iso-8859-1
#
# <!DOCTYPE HTML PUBLIC "-//IETF//DTD HTML 2.0//EN">
# <html><head>
# <title>400 Bad Request</title>
# </head><body>
# <h1>Bad Request</h1>
# <p>Your browser sent a request that this server could not understand.<br />
# </p>
# </body></html>
# <!DOCTYPE HTML PUBLIC "-//IETF//DTD HTML 2.0//EN">
# <html><head>
################################################################################

s.sendall(req)
resp = s.recv(4096).decode("utf-8")
s.close()
# print('Received:')
# print(resp)

check(resp, name, response, False)



# Example  coming out from Chrome
################################################################################
# GET /~wus2/cgi-bin/a.cgi HTTP/1.1
# Host: www-test.cs.umanitoba.ca
# Connection: keep-alive
# Cache-Control: max-age=0
# DNT: 1
# Upgrade-Insecure-Requests: 1
# User-Agent: Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_6) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/84.0.4147.89 Safari/537.36
# Accept: text/html,application/xhtml+xml,application/xml;q=0.9,image/webp,image/apng,*/*;q=0.8,application/signed-exchange;v=b3;q=0.9
# Accept-Encoding: gzip, deflate
# Accept-Language: en-CA,en;q=0.9,zh-CN;q=0.8,zh;q=0.7,ja-JP;q=0.6,ja;q=0.5
# Cookie: name=asdfsadf; accept=1; lastModified=0
################################################################################

# Debloated GET request
################################################################################
# GET /~wus2/cgi-bin/a.cgi HTTP/1.1
# Host: www-test.cs.umanitoba.ca
# Cookie: name=asdfsadf; accept=1; lastModified=0
################################################################################

# example server response
################################################################################
# HTTP/1.1 200 OK
# Date: Wed, 22 Jul 2020 08:29:25 GMT
# Server: Apache/2.4.6 (Scientific Linux) OpenSSL/1.0.2k-fips PHP/5.4.16 mod_wsgi/3.4 Python/2.7.5
# Transfer-Encoding: chunked
# Content-Type: text/html; charset=UTF-8
#
# 369
#
#     <html>
#     <body>
# <p style="color:green">Thanks for replying. Here's your previous response.</p>
#
#     <h1>You are Invited!</h1> 
#     details goes here
#     <br/><br/><br/>
#
#     <form action="/~wus2/cgi-bin/a.cgi" method="POST">
#         <label for="name">Your name:</label>
#         <input type="text" name="name" pattern=".*[a-zA-Z].*" title="At least 1 English letter" value=gdfsg autofocus required>
#         <br/>
#         <p2>Would you accept the invitation?</p2>
#         <br/>
#         <input type="radio" name="accept" value="1" required checked> Accept
#         <input type="radio" name="accept" value="0" > Decline
#         <br/>
#         <input type="submit" value="Submit">
#     </form>
#     <br/>
#    
#
#         <form action="/~wus2/cgi-bin/a.cgi" method="GET">
#         <button name="Anonymize" type="submit">Anonymize</button>
#         </form>
# 
# </body></html>
################################################################################


# build request string
req = b""
req += b"GET /~wus2/cgi-bin/a.cgi HTTP/1.1\r\n"
req += b"Host: www-test.cs.umanitoba.ca\r\n"
req += str.encode(f"Cookie: name={name}; accept={response}\r\n\r\n")


# create an INET, STREAMing socket
s = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
# now connect to the web server on port 80 - the normal http port
s.connect(("www-test.cs.umanitoba.ca", 80))
s.sendall(req)
resp = s.recv(4096).decode("utf-8")
s.close()
# print('Received:')
# print(resp)

check(resp, name, response, True)
print("Success.")

File: part1/scraper.c
#include <stdio.h>
#include <string.h>
#include <sys/socket.h>
#include <arpa/inet.h>
#include <assert.h>
#include <stdbool.h>
#include <ctype.h>

#include <netdb.h> // For getaddrinfo
#include <unistd.h> // for close
#include <stdlib.h> // for exit

///
/// @brief A test fuction to check if the server returned the thing we want.
///        It compares the result to check line by line, against a semi-hardcoded string.
/// @param toCheck string to check
/// @param name name of user
/// @param response response from user, either `"1"` for "yes", `"0"` for "no"
/// @return `true` if check passed, `false` if failed.
///         If `NDEBUG` is not defined(i.e. assersion is on), and a check failed, the function will be asserted instead.
bool check(char toCheck[], char* name, char* response)
{
    assert(toCheck);
    assert(name);
    assert(response);

    const char* CRLF = "\n";

    int nameCookieLength = 20 + strlen(name);
    char nameCookie[nameCookieLength];
    memset(nameCookie, '\0', nameCookieLength);
    strcat(nameCookie, "Set-Cookie: name=");
    strcat(nameCookie, name);
    strcat(nameCookie, "\r");

    char acceptCookie[24] = "Set-Cookie: accept=";
    strcat(acceptCookie, response);
    strcat(acceptCookie, "\r");

    int nameInputLength = 125+strlen(name);
    char nameInput[nameInputLength];
    memset(nameInput, '\0', nameInputLength);
    strcat(nameInput, "        <input type=\"text\" name=\"name\" pattern=\".*[a-zA-Z].*\" title=\"At least 1 English letter\" value=");
    strcat(nameInput, name);
    strcat(nameInput, " autofocus required>");

    char* accpet;
    char* decline;

    if (strcmp(response, "0") == 0)
    {
        accpet = "        <input type=\"radio\" name=\"accept\" value=\"1\" required > Accept";
        decline = "        <input type=\"radio\" name=\"accept\" value=\"0\" checked> Decline";
    }
    else if (strcmp(response, "1") == 0)
    {
        accpet = "        <input type=\"radio\" name=\"accept\" value=\"1\" required checked> Accept";
        decline = "        <input type=\"radio\" name=\"accept\" value=\"0\" > Decline";
    }
    else
    {
        assert(strcmp(response, "0") == 0);
        assert(strcmp(response, "1") == 0);
        return false;
    }

    assert(accpet);
    assert(decline);

    // reference have 25 lines
    const char* const REFERENCE[] = 
    {
        "HTTP/1.1 200 OK\r",
        // skip Date
        // skip Server
        nameCookie,
        acceptCookie,
        // skip transfer encoding
        "Content-Type: text/html; charset=UTF-8\r",
        // skip 3 lines
        "    <html>",
        "    <body>",
        "    <h1>You are Invited!</h1> ",
        "    details goes here",
        "    <br/><br/><br/>",
        "    <form action=\"/~wus2/cgi-bin/a.cgi\" method=\"POST\">",
        "        <label for=\"name\">Your name:</label>",
        nameInput,
        "        <br/>",
        "        <p2>Would you accept the invitation?</p2>",
        "        <br/>",
        accpet,
        decline,
        "        <br/>",
        "        <input type=\"submit\" value=\"Submit\">",
        "    </form>",
        "    <br/>",
        "        <form action=\"/~wus2/cgi-bin/a.cgi\" method=\"GET\">",
        "        <button name=\"Anonymize\" type=\"submit\">Anonymize</button>",
        "        </form>",
        "</body></html>",
    };
    
    #define skip     token = strtok(NULL, CRLF)
    #define refcmp(i) token = strtok(NULL, CRLF); /*printf("%d: %s &&  %s\n", i, token, REFERENCE[i]);*/ if (strcmp(token, REFERENCE[i]) != 0) {assert(strcmp(token, REFERENCE[i]) != 0); return false;}

    char* token = strtok(toCheck, CRLF);
    char* congrats;
    if (strcmp(token, REFERENCE[0]) != 0) {assert(strcmp(token, REFERENCE[0]) != 0); return false;}
    skip; // Date
    skip; // Server
    refcmp(1);
    refcmp(2);
    skip;
    refcmp(3);
    skip;
    skip;
    // skip;
    refcmp(4);
    refcmp(5);
    
    token = strtok(NULL, CRLF); // 12 
    if (strcmp(token, "<p style=\"color:green\">Your reply has been submitted!</p>") != 0 && 
        strcmp(token, "<p style=\"color:green\">Your reply has been updated!</p>") != 0)
    {
        //printf("congrat: %s \n", token);
        assert(strcmp(token, "<p style=\"color:green\">Your reply has been submitted!</p>") != 0 && 
                strcmp(token, "<p style=\"color:green\">Your reply has been updated!</p>") != 0);
        return false;
    }

    for (int i = 6; i <= 20; i++)
    {
        refcmp(i);
    }
    skip;
    refcmp(21);
    refcmp(22);
    refcmp(23);
    skip;
    refcmp(24);

    return true;
}

///
/// @brief A test fuction to check if the server returned the thing we want.
///        It compares the result to check line by line, against a semi-hardcoded string.
/// @param toCheck string to check
/// @param name name of user
/// @param response response from user, either `"1"` for "yes", `"0"` for "no"
/// @return `true` if check passed, `false` if failed.
///         If `NDEBUG` is not defined(i.e. assersion is on), and a check failed, the function will asserted instead.
bool check2(char toCheck[], char* name, char* response)
{
    assert(toCheck);
    assert(name);
    assert(response);

    const char* CRLF = "\n";

    int nameInputLength = 125+strlen(name);
    char nameInput[nameInputLength];
    memset(nameInput, '\0', nameInputLength);
    strcat(nameInput, "        <input type=\"text\" name=\"name\" pattern=\".*[a-zA-Z].*\" title=\"At least 1 English letter\" value=");
    strcat(nameInput, name);
    strcat(nameInput, " autofocus required>");

    char* accpet;
    char* decline;

    if (strcmp(response, "0") == 0)
    {
        accpet = "        <input type=\"radio\" name=\"accept\" value=\"1\" required > Accept";
        decline = "        <input type=\"radio\" name=\"accept\" value=\"0\" checked> Decline";
    }
    else if (strcmp(response, "1") == 0)
    {
        accpet = "        <input type=\"radio\" name=\"accept\" value=\"1\" required checked> Accept";
        decline = "        <input type=\"radio\" name=\"accept\" value=\"0\" > Decline";
    }
    else
    {
        assert(strcmp(response, "0") == 0);
        assert(strcmp(response, "1") == 0);
        return false;
    }

    assert(accpet);
    assert(decline);

    // reference have 24 lines
    const char* const REFERENCE[] = 
    {
        "HTTP/1.1 200 OK\r",
        // skip Date
        // skip Server
        // skip transfer encoding
        "Content-Type: text/html; charset=UTF-8\r",
        // skip 3 lines
        "    <html>",
        "    <body>",
        "<p style=\"color:green\">Thanks for replying. Here's your previous response.</p>",
        "    <h1>You are Invited!</h1> ",
        "    details goes here",
        "    <br/><br/><br/>",
        "    <form action=\"/~wus2/cgi-bin/a.cgi\" method=\"POST\">",
        "        <label for=\"name\">Your name:</label>",
        nameInput,
        "        <br/>",
        "        <p2>Would you accept the invitation?</p2>",
        "        <br/>",
        accpet,
        decline,
        "        <br/>",
        "        <input type=\"submit\" value=\"Submit\">",
        "    </form>",
        "    <br/>",
        "        <form action=\"/~wus2/cgi-bin/a.cgi\" method=\"GET\">",
        "        <button name=\"Anonymize\" type=\"submit\">Anonymize</button>",
        "        </form>",
        "</body></html>",
    };
    
    char* token = strtok(toCheck, CRLF);
    char* congrats;
    if (strcmp(token, REFERENCE[0]) != 0) {assert(strcmp(token, REFERENCE[0]) != 0); return false;}
    skip; // Date
    skip; // Server
    skip; // encoding
    refcmp(1);
    skip;
    skip;

    for (int i = 2; i <= 19; i++)
    {
        refcmp(i);
    }

    skip;
    refcmp(20);
    refcmp(21);
    refcmp(22);
    skip;
    refcmp(23);

    return true;
}

///
/// @brief Encapsulated web socket access.
/// @param request C-string holds the request
/// @param server_message out parameter, a C-string holds the response from server
/// @param server_message_length length of message to recevive
void sendNrequest(char* request, char* server_message, int server_message_length)
{
    int socket_desc;
    struct sockaddr_in server_addr;
    char address[100];

    struct addrinfo *result;
        
    // Create socket:
    socket_desc = socket(AF_INET, SOCK_STREAM, 0);
    
    if(socket_desc < 0){
        printf("Unable to create socket\n");
        exit(EXIT_FAILURE);
    }
    
    // printf("Socket created successfully\n");

    struct addrinfo hints;
    memset (&hints, 0, sizeof (hints));
    hints.ai_family = PF_UNSPEC;
    hints.ai_socktype = SOCK_STREAM;
    hints.ai_flags |= AI_CANONNAME;
    
    // get the ip of the page we want to scrape
    int out = getaddrinfo ("www-test.cs.umanitoba.ca", NULL, &hints, &result);
    // fail gracefully
    if (out != 0) {
        fprintf(stderr, "getaddrinfo: %s\n", gai_strerror(out));
        exit(EXIT_FAILURE);
    }
    
    // ai_addr is a struct sockaddr
    // so, we can just use that sin_addr
    struct sockaddr_in *serverDetails =  (struct sockaddr_in *)result->ai_addr;
    
    // Set port and IP the same as server-side:
    server_addr.sin_family = AF_INET;
    server_addr.sin_port = htons(80);
    //server_addr.sin_addr.s_addr = inet_addr("127.0.0.1");
    server_addr.sin_addr = serverDetails->sin_addr;
    
    // converts to octets
    // printf("Convert...\n");
    inet_ntop (server_addr.sin_family, &server_addr.sin_addr, address, 100);
    // printf("Connecting to %s\n", address);
    // Send connection request to server:
    if(connect(socket_desc, (struct sockaddr*)&server_addr, sizeof(server_addr)) < 0){
        printf("Unable to connect\n");
        exit(EXIT_FAILURE);
    }
    // printf("Connected with server successfully\n");

    // printf("Sending:\n########################\n%s\n#######################\n", request);
    // Send the message to server:
    // printf("Sending request, %lu bytes\n", strlen(request));
    if(send(socket_desc, request, strlen(request), 0) < 0){
        printf("Unable to send message\n");
        exit(EXIT_FAILURE);
    }
    
    // Receive the server's response:
    if(recv(socket_desc, server_message, server_message_length, 0) < 0){
        printf("Error while receiving server's msg\n");
        exit(EXIT_FAILURE);
    }
    // Close the socket:
    close(socket_desc);
}


bool containsAlpha(const char s[])
{
    unsigned char c;

    for (int i = 0; i < strlen(s); i++)
    {
        if (isalpha(s[i]))
        {
            return true;
        }
    }
    return false;
}

int main (int argc, char **argv)
{
    if (argc != 3 || (strcmp(argv[2], "yes") != 0 && strcmp(argv[2], "no") != 0) || !containsAlpha(argv[1]))
    {
        printf("Bad format\nExample usage:\n%s Rick yes\n%s Morty no\n", argv[0], argv[0]);
        return 1;
    }


    // char name[] = "idsjuafabfbadf";
    // char response[] = "0";
    char* name = argv[1];
    char* response = (strcmp(argv[2], "yes") == 0 ? "1" : "0");

    // prepare post request payload string
    int post_request_length = 15 + (int) strlen(name);
    char post_request[post_request_length];
    memset(post_request, '\0', post_request_length);
    snprintf(post_request, post_request_length, "name=%s&accept=%s", name, response);
    
    // prepare content length number
    int content_length_length = snprintf(NULL, 0, "%d", post_request_length) + 1;
    char content_length[content_length_length];
    memset(content_length, '\0', content_length_length);
    snprintf(content_length, content_length_length, "%d", post_request_length - 1);
    
    // build the whole post request string
    int request_length = 154 + (int) strlen(name);
    char* request = calloc(request_length, sizeof(char));

    strcat(request, "POST /~wus2/cgi-bin/a.cgi HTTP/1.1\r\n");
    strcat(request, "Host: www-test.cs.umanitoba.ca\r\n");
    strcat(request, "Content-Length: ");
    strcat(request, content_length);
    strcat(request, "\r\n");
    strcat(request, "Content-Type: application/x-www-form-urlencoded\r\n\r\n");
    strcat(request, post_request);

    char* server_message = calloc(2000, sizeof(char));
    sendNrequest(request, server_message, 2000);
    // printf("Server's response:\n%s", server_message);
    if (!check(server_message, name, response))
    {
        printf("%s\n", "Test 1 failed.");
        free(request);
        free(server_message);
        return(2);
    }

    free(request);
    free(server_message);


    // prepare the GET request
    request_length = 95 + (int) strlen(name);
    request = calloc(request_length, sizeof(char));

    // prepare cookie string 
    int cookie_string_length = 24 + (int) strlen(name);
    char cookie_string[cookie_string_length];
    memset(cookie_string, '\0', cookie_string_length);
    snprintf(cookie_string, cookie_string_length, "Cookie: name=%s; accept=%s", name, response);

    // build the GET request string
    strcat(request, "GET /~wus2/cgi-bin/a.cgi HTTP/1.1\r\n");
    strcat(request, "Host: www-test.cs.umanitoba.ca\r\n");
    strcat(request, cookie_string);
    strcat(request, "\r\n\r\n");

    server_message = calloc(2000, sizeof(char));
    sendNrequest(request, server_message, 2000);
    // printf("Server's response:\n%s", server_message);
    if (!check2(server_message, name, response))
    {
        printf("%s\n", "Test 2 failed.");
        free(request);
        free(server_message);
        return(2);
    }

    free(request);
    free(server_message);

    printf("Success.\n");
    return 0;
}
